### ISSUE TITLE
learning rate not restored properly when using a CosineAnnealingLR scheduler

### ISSUE THREAD

**Issue by quinor** [PyTorchLightning/pytorch-lightning#5455]
## 🐛 Bug

When resuming training from a saved checkpoint, learning rate is not restored. It causes the learning rate to follow incorrect curve. The issue is most prominent when using a multiplicative LR scheduler (ie. `torch.optim.lr_scheduler.CosineAnnealingLR`, instead of assigning a new value it multiplies the old one by a specific factor), but I've encountered similar behavior with other schedulers too. An example below:

![image](https://user-images.githubusercontent.com/3879392/104151518-7f16a380-53dd-11eb-8723-619573e5a62a.png)

### To Reproduce

Create a model with `torch.optim.lr_scheduler.CosineAnnealingLR`. Train for several epochs, checkpoint, resume training from the checkpoint. The learning rate should then follow similar (incorrect) curve as in the example above.

**Comment by tchaton** [comment]
Hey @quinor,

Would you mind reproducing the bug with [bug_report_model](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/bug_report_model.py), it would save us some time as we prioritise bugs with code.

Best regards,
T.C

**Comment by quinor** [comment]
I'm very short with time recently. I'll try to work on that, but it might be a few weeks until I find a time to reproduce it outside my research code.

**Comment by AndersJing** [comment]
I also find this bug and the scheduler I use is `torch.optim.lr_scheduler.CyclicLR`. 

When using the `resume_from_checkpoint` to restore the training process from a checkpoint, the learning rate always starts from the initial learning rate.

**Comment by edenlightning** [comment]
@akihironitta mind taking a look?

**Comment by akihironitta** [comment]
@quinor @AndersJing My below code successfully resumes from the checkpoint with `CosineAnnealingLR` used. In order to investigate further, would you mind providing us with your script?

![tmp](https://user-images.githubusercontent.com/20610905/106534305-27c8a680-6537-11eb-94a4-ef23aeb0c1b5.png)

<details>
<summary>my code</summary>

```python
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torch.utils.data import DataLoader, random_split
from torchvision import transforms
from torchvision.datasets import MNIST

import pytorch_lightning as pl
from pytorch_lightning.loggers import TensorBoardLogger
from pytorch_lightning.callbacks import LearningRateMonitor, ModelCheckpoint

pl.seed_everything(42)


class LightningMNISTClassifier(pl.LightningModule):
    def __init__(self):
        super(LightningMNISTClassifier, self).__init__()
        self.layer_1 = nn.Linear(28 * 28, 128)
        self.layer_2 = nn.Linear(128, 256)
        self.layer_3 = nn.Linear(256, 10)

    def forward(self, x):
        batch_size, channels, width, height = x.size()
        x = x.view(batch_size, -1)
        x = self.layer_1(x)
        x = torch.relu(x)
        x = self.layer_2(x)
        x = torch.relu(x)
        x = self.layer_3(x)
        x = torch.log_softmax(x, dim=1)
        return x

    def prepare_data(self):
        transform = transforms.Compose(
            [transforms.ToTensor(), transforms.Normalize((0.1307,), (0.3081,))]
        )
        mnist_train = MNIST("datasets", train=True, download=True, transform=transform)
        self.mnist_train, self.mnist_val = random_split(
            mnist_train, [55000, 5000], generator=torch.Generator().manual_seed(42)
        )

    def train_dataloader(self):
        return DataLoader(self.mnist_train, batch_size=1024, num_workers=0)

    def configure_optimizers(self):
        optimizer = optim.SGD(self.parameters(), lr=0.1)
        lr_scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=10)
        # lr_scheduler = torch.optim.lr_scheduler.CyclicLR(optimizer, base_lr=0.01, max_lr=0.1)
        return [optimizer], [lr_scheduler]

    def training_step(self, train_batch, batch_idx):
        x, y = train_batch
        logits = self.forward(x)
        loss = F.nll_loss(logits, y)
        self.log("loss", loss)
        return {"loss": loss}


def main1():
    model_checkpoint = ModelCheckpoint(save_last=True, monitor="loss")
    lr_monitor = LearningRateMonitor(logging_interval="step")
    callbacks = [model_checkpoint, lr_monitor]
    logger = TensorBoardLogger('logs_5455/')

    model = LightningMNISTClassifier()

    trainer = pl.Trainer(
        log_every_n_steps=1,
        max_epochs=3,
        logger=logger,
        callbacks=callbacks,
    )
    trainer.fit(model)

def main2():
    lr_monitor = LearningRateMonitor(logging_interval="step")
    callbacks = [lr_monitor]
    logger = TensorBoardLogger('logs_5455/')

    model = LightningMNISTClassifier()

    trainer = pl.Trainer(
        resume_from_checkpoint="logs_5455/default/version_0/checkpoints/last.ckpt",
        log_every_n_steps=1,
        max_epochs=5,
        logger=logger,
        callbacks=callbacks,
    )
    trainer.fit(model)

if __name__ == "__main__":
    main1()
    main2()
```

</details>

<details>
<summary>my env</summary>

```console
$ wget https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py
$ python collect_env_details.py
* CUDA:
        - GPU:
        - available:         False
        - version:           None
* Packages:
        - numpy:             1.19.5
        - pyTorch_debug:     False
        - pyTorch_version:   1.7.1+cpu
        - pytorch-lightning: 1.1.6
        - tqdm:              4.56.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         
        - python:            3.8.5
        - version:           #1 SMP Debian 4.19.160-2 (2020-11-28)
```

<details>

**Comment by Alexfinkelshtein** [comment]
> ```
> $ wget https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py
> $ python collect_env_details.py
> ```

Hi I am facing a similar issue and found this post, than tried running the code you shared on colab and I couldn't reproduce the behavior you show for the learning rate
I see that the scheduling and learning rates start from the beginning as if the state wasn't restored
![pl_issue2](https://user-images.githubusercontent.com/16181467/107496079-2af41000-6b91-11eb-98f4-b158438f3df6.png)

(by the way it seems like the script is wrong with respect to GPU availability)
* CUDA:
	- GPU:
	- available:         False
	- version:           10.2
* Packages:
	- numpy:             1.19.5
	- pyTorch_debug:     False
	- pyTorch_version:   1.7.1
	- pytorch-lightning: 1.1.6
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.6.9
	- version:           #1 SMP Thu Jul 23 08:00:38 PDT 2020

**Comment by akihironitta** [comment]
> tried running the code you shared on colab and I couldn't reproduce the behavior you show for the learning rate

@Alexfinkelshtein Hi, thank you for your report. I might be missing, but which script are you referring to?

> (by the way it seems like the script is wrong with respect to GPU availability)

You mean the script `collect_env_details.py` outputs `available: False` where you actually have GPUs? If so, the script simply prints `torch.cuda.is_available(),`, so the issue is not Lightning or PyTorch side, I suppose.

**Comment by Alexfinkelshtein** [comment]
thanks, I refer to this:
`import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torch.utils.data import DataLoader, random_split
from torchvision import transforms
from torchvision.datasets import MNIST

import pytorch_lightning as pl
from pytorch_lightning.loggers import TensorBoardLogger
from pytorch_lightning.callbacks import LearningRateMonitor, ModelCheckpoint

pl.seed_everything(42)


class LightningMNISTClassifier(pl.LightningModule):
    def __init__(self):
        super(LightningMNISTClassifier, self).__init__()
        self.layer_1 = nn.Linear(28 * 28, 128)
        self.layer_2 = nn.Linear(128, 256)
        self.layer_3 = nn.Linear(256, 10)

    def forward(self, x):
        batch_size, channels, width, height = x.size()
        x = x.view(batch_size, -1)
        x = self.layer_1(x)
        x = torch.relu(x)
        x = self.layer_2(x)
        x = torch.relu(x)
        x = self.layer_3(x)
        x = torch.log_softmax(x, dim=1)
        return x

    def prepare_data(self):
        transform = transforms.Compose(
            [transforms.ToTensor(), transforms.Normalize((0.1307,), (0.3081,))]
        )
        mnist_train = MNIST("datasets", train=True, download=True, transform=transform)
        self.mnist_train, self.mnist_val = random_split(
            mnist_train, [55000, 5000], generator=torch.Generator().manual_seed(42)
        )

    def train_dataloader(self):
        return DataLoader(self.mnist_train, batch_size=1024, num_workers=0)

    def configure_optimizers(self):
        optimizer = optim.SGD(self.parameters(), lr=0.1)
        lr_scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=10)
        # lr_scheduler = torch.optim.lr_scheduler.CyclicLR(optimizer, base_lr=0.01, max_lr=0.1)
        return [optimizer], [lr_scheduler]

    def training_step(self, train_batch, batch_idx):
        x, y = train_batch
        logits = self.forward(x)
        loss = F.nll_loss(logits, y)
        self.log("loss", loss)
        return {"loss": loss}


def main1():
    model_checkpoint = ModelCheckpoint(save_last=True, monitor="loss")
    lr_monitor = LearningRateMonitor(logging_interval="step")
    callbacks = [model_checkpoint, lr_monitor]
    logger = TensorBoardLogger('logs_5455/')

    model = LightningMNISTClassifier()

    trainer = pl.Trainer(
        log_every_n_steps=1,
        max_epochs=3,
        logger=logger,
        callbacks=callbacks,
    )
    trainer.fit(model)

def main2():
    lr_monitor = LearningRateMonitor(logging_interval="step")
    callbacks = [lr_monitor]
    logger = TensorBoardLogger('logs_5455/')

    model = LightningMNISTClassifier()

    trainer = pl.Trainer(
        resume_from_checkpoint="logs_5455/default/version_0/checkpoints/last.ckpt",
        log_every_n_steps=1,
        max_epochs=5,
        logger=logger,
        callbacks=callbacks,
    )
    trainer.fit(model)

if __name__ == "__main__":
    main1()
    main2()`

**Comment by akihironitta** [comment]
@Alexfinkelshtein @quinor @AndersJing I tried again with the BoringModel [here on Google Colab](https://colab.research.google.com/drive/1uUMhbdtXaNfSbJXrBaCmplrhY-IuLjpW?usp=sharing), but I couldn't reproduce the bug you're reporting. Would you be able to reproduce it with [**BoringModel**](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing) and post here?

**Comment by Alexfinkelshtein** [comment]
> @Alexfinkelshtein @quinor @AndersJing I tried again with the BoringModel [here on Google Colab](https://colab.research.google.com/drive/1uUMhbdtXaNfSbJXrBaCmplrhY-IuLjpW?usp=sharing), but I couldn't reproduce the bug you're reporting. Would you be able to reproduce it with [**BoringModel**](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing) and post here?

@akihironitta  I ran the code again and I can confirm that I had a mistake, I didn't restart  the runtime when testing different pytorch-lightning versions. Now I can confirm: 
1.1.6 works correctly
1.1.0 doesn't pick up scheduling from the right place when checkpoint restored
as far as I understand the issue is resolved
thanks

**Comment by quinor** [comment]
I can confirm I had the issue on most recent 1.1.4. I don't have time to check it now with 1.1.6, but I guess it's OK for me to close the issue and just reopen it if I can get the reproduce on new lightning.

**Comment by akihironitta** [comment]
@Alexfinkelshtein @quinor Thank you for your confirmation. Feel free to reopen this issue if you experience the bug again (with a reproducible script would be very helpful).

Could anyone from core close this issue as there seem to be nothing to investigate further?

**Comment by tchaton** [comment]
Closing this issue as it seemed to be resolved. Feel free to re-open it.

**Comment by AndersJing** [comment]
I can confirm the issue has been resolved on v1.2.1 when I use MultiStepLR.

**Comment by gahdritz** [comment]
This bug is back in 1.6.0 (at least). Here's a slightly updated version of @akihironitta 's [BoringModel Colab](https://colab.research.google.com/drive/1JUEZYYbgZnWfoN-v_4XR4AAVOBTV1Phw?usp=sharing). I just removed a couple of unnecessary imports that don't work anymore.

**Comment by woshiyyya** [comment]
Confirm that it has been resolved on v2.1.3 with `torch.optim.lr_scheduler.LambdaLR`
